File: dsa-problems/Greedy/solutions/cpp/GRD-014-festival-bandwidth-split.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxStages(int n, long long B, vector<long long>& bandwidths) {
        sort(bandwidths.begin(), bandwidths.end());
        
        long long currentSum = 0;
        int count = 0;
        
        for (long long b : bandwidths) {
            if (currentSum + b <= B) {
                currentSum += b;
                count++;
            } else {
                break;
            }
        }
        
        return count;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    long long B;
    if (!(cin >> n >> B)) return 0;

    vector<long long> bandwidths(n);
    for (int i = 0; i < n; i++) {
        cin >> bandwidths[i];
    }

    Solution solution;
    cout << solution.maxStages(n, B, bandwidths) << "\n";

    return 0;
}

File: dsa-problems/Greedy/solutions/java/GRD-014-festival-bandwidth-split.java
import java.util.*;

class Solution {
    public int maxStages(int n, long B, long[] bandwidths) {
        Arrays.sort(bandwidths);
        
        long currentSum = 0;
        int count = 0;
        
        for (long b : bandwidths) {
            if (currentSum + b <= B) {
                currentSum += b;
                count++;
            } else {
                break;
            }
        }
        
        return count;
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        
        int n = sc.nextInt();
        long B = sc.nextLong();
        
        long[] bandwidths = new long[n];
        for (int i = 0; i < n; i++) {
            bandwidths[i] = sc.nextLong();
        }
        
        Solution solution = new Solution();
        System.out.println(solution.maxStages(n, B, bandwidths));
        sc.close();
    }
}
